File: src/DocumentProcessor.java
public class DocumentProcessor {

    private static final Stemmer stemmer;
    private String data;

    static {
        stemmer = new Stemmer();
    }


    public DocumentProcessor(String data) {
        this.data = data;
    }


    private void toLowerCase() {
        if (data == null) return;
        data = data.toLowerCase();
    }

    private void removeSigns() {
        if (data == null) return;
        removeDashes();
        removeNonAlphaNumerics();
    }

    private void removeDashes() {
        data = data.replaceAll("[-?]+", " ");
    }

    private void removeNonAlphaNumerics() {
        data = data.replaceAll("[^a-zA-Z0-9\\s]+", " ");
    }

    private String[] toStemmedSplit() {
        if (data == null) return new String[0];
        String[] splitData = splitData();
        return stemSplitData(splitData);

    }

    private String[] stemSplitData(String[] splitData) {
        for (int i = 0; i < splitData.length; i++) {
            splitData[i] = stemmer.stemWord(splitData[i]);
        }
        return splitData;
    }

    private String[] splitData() {
        return data.split("\\s+");
    }

    public String[] getNormalizedWords() {
        toLowerCase();
        removeSigns();
        return toStemmedSplit();
    }

}

File: searchProject_csharp/SampleLibrary/SampleLibrary/DocumentProcessor.cs
using System.Text.RegularExpressions;
namespace SampleLibrary
{
    public class DocumentProcessor
    {
        private readonly string data;
        private static readonly Stemmer stemmer = new Stemmer();
        public DocumentProcessor(string data)
        {
            this.data = data.ToLower() ?? throw new Exception("data is null");
        }
        public string[] getNormalizedWords()
        {
            return toStemSplit(removeSigns());
        }

        private string removeSigns()
        {
            return removeNonNumericalSigns(removeDashes());
        }

        private string removeDashes()
        {
            var regex = new Regex("[-?]+");
            return regex.Replace(data," ");
        }

        private string removeNonNumericalSigns(string temp)
        {
            var regex = new Regex("[^a-zA-Z0-9\\s]+");
            return regex.Replace(temp," ");
        }
        
        private string[] toStemSplit(string temp)
        {
            var splittedData = temp.Split(" ");
            return stemSplitData(splittedData);
        }

        private string[] stemSplitData(string[] splittedData)
        {
            return splittedData.Select(p => stemmer.StemWord(p)).ToArray();
        }

    }
}